### PR TITLE
RFC: Meson cleanups

### PR DESCRIPTION
**Pull request by woju**
## Description of the changes 

- use `type: 'feature'` for enabled/disabled options
- remove `enabled_` prefix from some of the option-derived variables
- use `/` instead of `join_paths()`
- use `prefix` variable instead of `get_option('prefix')`

## RFC

WRT `option(type: 'feature')`, there are three style variants, which are possible now.

variant 1: variable is boolean
```
sgx = get_option('sgx').enabled()
libfoo_dep = dependency('foo', required: sgx)
if sgx ...
```

variant 2: variable is a [feature](https://mesonbuild.com/Reference-manual_returned_feature.html)
```
sgx = get_option('sgx')
libfoo_dep = dependency('foo', required: sgx)
if sgx.enabled() ...
```

variant 3: no intermediate variable, call `get_option()` every time
```
libfoo_dep = dependency('foo', required: get_option('sgx'))
if get_option('sgx').enabled() ...
```

IDK which variant do we need, I did variant 1 here, because it was closest, and removed `enabled_` for consistency, but if `enabled_` was somehow needed, I think I can do variant 2 instead.

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)

### CODE REVIEW

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 33 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KbGZF2UJ1FLtk4h6o:-N3KbGZG-3V6tvK6mmN2:bmxqwlx) (no related file):*
Since which version features are enabled? I could not find it in the docs. You stated `for compatibility with older meson versions which did not support feature options`, I'm just curious which version was that

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KcBUk0NOb5tC_N_vO:-N3KcBUk0NOb5tC_N_vP:bk0wcm2) (no related file):*
AD RFC: if the 3rd option doesn't look to ugly I would prefer it. The main reason is with variable names: `sgx` tells us nothing about what it is. `sgx.enabled()` is fine, but you have to know that `sgx` is a feature. `get_option('sgx').enabled() ` clearly shows what's going on.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Ked_-3DEVhg18u7NV:-N3Ked_-3DEVhg18u7NW:b-dj97o5) (no related file):*
I dislike the `get_option('prefix')` -> `prefix` change. `prefix` is kind of vague, no idea prefix to what...
Maybe `project_prefix` insread?

___
___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kb81r7KDDiBsVkZc4):*
https://gramine.readthedocs.io/en/latest/devel/contributing.html#reviewing-guidelines point 2 :)
___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KbefK4ZX2iV16TM_r):*
ditto
___
*[`meson.build` line 28 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kbwu60uzJRFzPERIh:-N3Kbwu60uzJRFzPERIi:b18syq2) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/meson.build#L28)):*
> ```Meson
> syslibdir = get_option('syslibdir')
> if syslibdir == ''
>     syslibdir = '/' / get_option('libdir')
> ```

Out of curiosity: will this produce `/libdir` or `//libdir`?
___
*[`meson.build` line 281 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KdHKAEKYjNCC4VWrv:-N3KdHKB3eCCq6ybY_Jv:b2ni78k) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/meson.build#L281)):*
> ```Meson
>     endif
> 
>     vtune_dep = []
> ```

lol
___
*[`meson_options.txt` line 1 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KdTTS9YxWkVXV_1fJ:-N3KdTTS9YxWkVXV_1fK:bzga4il) ([raw file](https://github.com/gramineproject/gramine/blob/fe64d8657578261c6a87e454b15dccc5e562c271/meson_options.txt#L1)):*
> ```Meson
> # TODO: after deprecating 18.04/bionic, change these to type: 'feature'
> ```

But wait, we did not deprecate Ubuntu 18.04 (yet)

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)* status: 25 of 33 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski and @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KbGZF2UJ1FLtk4h6o:-N3Kid1yElUo0HV0QMvb:bxfjr0c) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Since which version features are enabled? I could not find it in the docs. You stated `for compatibility with older meson versions which did not support feature options`, I'm just curious which version was that
</blockquote></details>

They appeared between 0.45 (version in Ubuntu 18.04) which I originally targeted, and 0.55 which we have now. The exact version was 0.47: https://mesonbuild.com/Release-notes-for-0-47-0.html#new-type-of-build-option-for-features.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KcBUk0NOb5tC_N_vO:-N3Kik-w15C1Kmu8yr-a:b-6lwzgn) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

AD RFC: if the 3rd option doesn't look to ugly I would prefer it. The main reason is with variable names: `sgx` tells us nothing about what it is. `sgx.enabled()` is fine, but you have to know that `sgx` is a feature. `get_option('sgx').enabled() ` clearly shows what's going on.
</blockquote></details>

It's not that it's ugly, it's slighly more writing, but it can be done.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Ked_-3DEVhg18u7NV:-N3KimgI8Lut84sUXL5r:b-egai6k) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

I dislike the `get_option('prefix')` -> `prefix` change. `prefix` is kind of vague, no idea prefix to what...
Maybe `project_prefix` insread?
</blockquote></details>

I can do like variant 3, `get_option('prefix')` everywhere. WDYT?

___
___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kb81r7KDDiBsVkZc4):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

https://gramine.readthedocs.io/en/latest/devel/contributing.html#reviewing-guidelines point 2 :)
</blockquote></details>

@mkow told me to not write `[*]` in front, because something something candle.
___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KbefK4ZX2iV16TM_r):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto `[*]`
___
*[`meson.build` line 28 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kbwu60uzJRFzPERIh:-N3KkNFv7MpWI_aN0Olz:bphmhqa) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/meson.build#L28)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Out of curiosity: will this produce `/libdir` or `//libdir`?
</blockquote></details>

`'/lib'` most likely. `libdir` is virtually always relative (to `prefix`), usually `'lib'` or `'lib/x86_64-linux-gnu'`, and `prefix` is `'/usr'` or `'/usr/local'` by default, but can vary wildly. The point of this variable is to always be system-wide `'/lib'` even if `prefix` is `'/home/user/.local'`.
___
*[`meson.build` line 281 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KdHKAEKYjNCC4VWrv:-N3KlKDMEHyurCHVzzhR:b-if3xd8) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/meson.build#L281)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

lol
</blockquote></details>

yeah, this is overridden below `if sgx`.
___
*[`meson_options.txt` line 1 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KdTTS9YxWkVXV_1fJ:-N3KlQ-41cqlHlqSBata:bfx4rh) ([raw file](https://github.com/gramineproject/gramine/blob/fe64d8657578261c6a87e454b15dccc5e562c271/meson_options.txt#L1)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

But wait, we did not deprecate Ubuntu 18.04 (yet)
</blockquote></details>

This comment was written when I was targeting 0.45, but during discussion we decided to require meson 0.55 because of `subprojects/packagefiles` support. This meant that we didn't ever support meson from Ubuntu 18.04. But this `combo` hack wasn't rewritten either, so it remained as it were (same for `/` vs `join_paths()`, we could have done this earlier).

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 3 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)* status: 28 of 33 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski, @mkow, and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Ked_-3DEVhg18u7NV:-N3KtwEz6TH1r-d0BgwT:b405cg7) (no related file):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

I can do like variant 3, `get_option('prefix')` everywhere. WDYT?
</blockquote></details>

Fine with me. Having a global variable is also fine, as long as the name is not ambiguous. Up to you

___
___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kb81r7KDDiBsVkZc4):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

@mkow told me to not write `[*]` in front, because something something candle.
</blockquote></details>

What `[*]`?? `[Meson]` is what we use here (instead of `meson:`)
___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KuNS2Dw8UDlR1wDKA):*
ditto
___
*[`-- commits` line 23 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KuO7K7B_WW_oM7Fgr):*
ditto
___
*[`meson.build` line 28 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Kbwu60uzJRFzPERIh:-N3KuZ5N6m4VRofZoBMK:b-gqk0nv) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/meson.build#L28)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

`'/lib'` most likely. `libdir` is virtually always relative (to `prefix`), usually `'lib'` or `'lib/x86_64-linux-gnu'`, and `prefix` is `'/usr'` or `'/usr/local'` by default, but can vary wildly. The point of this variable is to always be system-wide `'/lib'` even if `prefix` is `'/home/user/.local'`.
</blockquote></details>

? I was talking about `/` vs `//`. Anyway, not important.
___
*[`common/proto/meson.build` line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KwKIw0iUAbnZm75NS:-N3KwKIw0iUAbnZm75NT:bnm38yb) ([raw file](https://github.com/gramineproject/gramine/blob/4d4db57a91b210e3ebbf921aca76970ebe80acc5/common/proto/meson.build#L2)):*
> ```Meson
> 
> # TODO after we drop Ubuntu 18.04 (ninja 1.8):
> ```

But we still install ninja from `apt`, this never changed (unlike `meson`)
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 6 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3KwaNB0diiCx3h7GFP:-N3KwaNB0diiCx3h7GFQ:b-5x2ssq) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/LibOS/shim/test/abi/x86_64/meson.build#L6)):*
> ```Meson
> )
> 
> testparams = {
> ```

If you remove global `tests` (which I as I understand you will) this can be reverted

**Comment by eli-schwartz** [comment]
> WRT `option(type: 'feature')`, there are three style variants, which are possible now.

The originally intended way to use these feature options in Meson is like this:

```meson
libfoo_dep = dependency('foo', required: get_option('sgx'))
if libfoo_dep.found() ...
```

Because the value of the sgx option isn't actually important for the state of your build. You don't care whether it's enabled, you care whether libfoo has been found and added to your build state.

If sgx is auto, and libfoo_dep is *found*, checking `sgx.enabled()` will say "nope, not enabled, ignore libfoo_dep and don't do any of this stuff".
If:
- sgx is disabled, then libfoo_dep is guaranteed to not be found,
- likewise if sgx is auto, and libfoo_dep happens to not be found
then in both cases, `sgx.enabled()` and `libfoo_dep.found()` are going to return the same results

If you're just going to logic gate off of sgx.enabled() then in fact you are only using a boolean state anyway. You could just as well stick with booleans. All you're really doing is having true/false options but *spelling* them "enabled" / "disabled" and also (confusingly) having "auto" act as "disabled".

> Since which version features are enabled? I could not find it in the docs. You stated `for compatibility with older meson versions which did not support feature options`, I'm just curious which version was that

The most important thing about which version this was added in, is whether your current minimum requirements allow it. You specify a minimum `meson_version` of 0.55.0, so Meson will warn you if you use functionality from a too-new version. 

(In this case, it is available since 0.47.0 as documented in the option types documentation at https://mesonbuild.com/Build-options.html#features.)

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 6 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)* status: all files reviewed, 13 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski, @mkow, and @woju)

___
*[`meson.build` line 49 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NpwSIDM7Q_rFOJyNB:-N3NpwSIDM7Q_rFOJyNC:b-zgbrdz) ([raw file](https://github.com/gramineproject/gramine/blob/fe64d8657578261c6a87e454b15dccc5e562c271/meson.build#L49)):*
> ```Meson
> objcopy = find_program('objcopy')
> 
> # TODO: after deprecating 18.04/bionic, update this to import('python')
> ```

But we haven't yet deprecated 18.04?

Is this PR waiting for another PR that will move our CI/docs/rest from 18.04 to 22.04?

How will it work on CentOS/Fedora?
___
*[`meson.build` line 272 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NqVqR9nFDme15dyi7:-N3NqVqR9nFDme15dyi8:b-eh8sl3) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/meson.build#L272)):*
> ```Meson
>     endif
>     if not cjson_dep.found()
>         # Ubuntu 18.04 does not have cjson at all
> ```

If we're deprecating 18.04 in this PR, this can also go away
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 6 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Nr11s5XW_l72Z70xA:-N3Nr11s5XW_l72Z70xB:b-h883kd) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/LibOS/shim/test/abi/x86_64/meson.build#L6)):*
> ```Meson
> )
> 
> testparams = {
> ```

Also, why such a weird name? These are separate programs, not params. So maybe `testcases` or `testprogs`?
___
*[`LibOS/shim/test/fs/meson.build` line 16 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NrASv9eWT_Uuelx4Z:-N3NrASwDVzyooS1_gdf:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/LibOS/shim/test/fs/meson.build#L16)):*
> ```Meson
> )
> 
> testparams = {
> ```

ditto
___
*[`LibOS/shim/test/regression/meson.build` line 1 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NrC1G75Awlac6Kg94:-N3NrC1G75Awlac6Kg95:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/LibOS/shim/test/regression/meson.build#L1)):*
> ```Meson
> testparams = {
> ```

ditto
___
*[`Pal/regression/meson.build` line 11 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NrHdC1SwNk4DfRdAA:-N3NrHdC1SwNk4DfRdAB:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/Pal/regression/meson.build#L11)):*
> ```Meson
> endif
> 
> testparams = {
> ```

ditto

**Review by boryspoplawski — COMMENTED** [review]
Most of these options (that we use) are indeed simple boolean switches (actually all of them are).

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619#-:-N3P1bRP3FLSrTU4TNRG:b-2m1ik8)* status: all files reviewed, 13 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski, @dimakuv, @mkow, and @woju)

___
*[`meson.build` line 49 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NpwSIDM7Q_rFOJyNB:-N3OqmRa0-maDfx-uykw:b-o214t) ([raw file](https://github.com/gramineproject/gramine/blob/fe64d8657578261c6a87e454b15dccc5e562c271/meson.build#L49)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

But we haven't yet deprecated 18.04?

Is this PR waiting for another PR that will move our CI/docs/rest from 18.04 to 22.04?

How will it work on CentOS/Fedora?
</blockquote></details>

There is a closed discussion about it somewhere in this PR, I had the same question :)

TL;DR: we are not deprecating 18.04, but this comment was misleading - it depended on old meson version, which is the one you get in Ubuntu 18.04, but we increased the meson minimal version anyway (and install it from pip there).
___
*[`meson.build` line 272 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3NqVqR9nFDme15dyi7:-N3OrEOM4fkurGv76iUc:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/meson.build#L272)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

If we're deprecating 18.04 in this PR, this can also go away
</blockquote></details>

ditto
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 6 at r2](https://reviewable.io/reviews/gramineproject/gramine/619#-N3Nr11s5XW_l72Z70xA:-N3OrKh5Atzor3CnDOGM:bpd8fqw) ([raw file](https://github.com/gramineproject/gramine/blob/3c9d2238ca4de0fac81d414ff530c1835fdc34a1/LibOS/shim/test/abi/x86_64/meson.build#L6)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Also, why such a weird name? These are separate programs, not params. So maybe `testcases` or `testprogs`?
</blockquote></details>

I would just revert this and not worry

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/619)* status: all files reviewed, 14 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski, @mkow, and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/619#-NYb9qoeAAWxBzP1voxT:-NYb9qoeAAWxBzP1voxU:bd7kft1) (no related file):*
@woju Just FYI. I recreated some of the commits/changes that you had in this PR. This was started as an independent effort (as response to #1400), and then I found out that you did similar changes in this PR. However, since this PR had no movement for a year already, I decided it will make sense to create a bunch of smaller PRs that recreation different sub-items from this PR.

> -   use `type: 'feature'` for enabled/disabled options

Done in https://github.com/gramineproject/gramine/pull/1414.

> -   remove `enabled_` prefix from some of the option-derived variables

I don't know why this change was important. I didn't recreate this change, looks useless to me.

> -   use `/` instead of `join_paths()`

This was done long time ago in https://github.com/gramineproject/gramine/commit/50712e97fc53d1c80e1e86b03736880c06eade47

> -   use `prefix` variable instead of `get_option('prefix')`

Done in https://github.com/gramineproject/gramine/pull/1415. Actually, not this solution but solution 3 as discussed in another comment (see top-level comment here between Woju and Borys).

---

The only important part form this PR not yet recreated is this commit: `meson: use multiple outputs for aesm_proto_ch custom_target()`.

___
